controller/api/destination/endpoint_translator.go

```diff
@@ -480,6 +490,74 @@ func toAddr(address watcher.Address) (*net.TcpAddress, error) {
 	}, nil
 }
 
+func createWeightedAddrForExternalWorkload(
+	address watcher.Address,
+	opaquePorts map[uint32]struct{},
+) (*pb.WeightedAddr, error) {
+	tcpAddr, err := toAddr(address)
+	if err != nil {
+		return nil, err
+	}
+
+	weightedAddr := pb.WeightedAddr{
+		Addr:         tcpAddr,
+		Weight:       defaultWeight,
+		MetricLabels: map[string]string{},
+	}
+
+	weightedAddr.MetricLabels = pkgK8s.GetExternalWorkloadLabels(address.OwnerKind, address.OwnerName, address.ExternalWorkload)
+	// If the address is not backed by an ExternalWorkload, there is no additional metadata
+	// to add.
+	if address.ExternalWorkload == nil {
+		return &weightedAddr, nil
+	}
+
+	weightedAddr.ProtocolHint = &pb.ProtocolHint{}
+	_, opaquePort := opaquePorts[address.Port]
+	// If address is set as opaque by a Server, or its port is set as
+	// opaque by annotation or default value, then set the hinted protocol to
+	// Opaque.
+	if address.OpaqueProtocol || opaquePort {
+		weightedAddr.ProtocolHint.Protocol = &pb.ProtocolHint_Opaque_{
+			Opaque: &pb.ProtocolHint_Opaque{},
+		}
+
+		port, err := getInboundPortFromExternalWorkload(&address.ExternalWorkload.Spec)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read inbound port: %w", err)
+		}
+		weightedAddr.ProtocolHint.OpaqueTransport = &pb.ProtocolHint_OpaqueTransport{
+			InboundPort: port,
```

Comment:
There's an interesting and subtle difference between this and our main opaque ports code (in `createWeightedAddr`). In `createWeightedAddr`, we attempt to get the proxy's inbound port and if it is missing (see https://github.com/linkerd/linkerd2/blob/main/controller/api/destination/endpoint_translator.go#L593-L611) then we return an error.

If the inbound port is missing, we can reasonably say there's no proxy env in the configured pod spec, so no proxy listening. However in this scenario, if the port is missing it does not imply the proxy isn't listening, we kind of assume that all external workloads would be meshed (since mesh tls settings are required). Some questions I'd have:

1. Should we default to `4143` (or whatever the inbound proxy port is) and log an error? Allows us to keep the endpoint in the pool and deal with misconfiguration through defaulting (in case users forgot to document the port).
2. Should we instead allow traffic to the proxy without an opaque hint? We should make it obvious that an error occurred and send traffic without re-mapping the port. Clients will run into protocol detect timeouts but we ensure the endpoint is kept in the pool.

I also think the behaviour we see now is good, just some ideas on how we could handle it.

Reply:
I agree that this is analogous to the case where a Pod has a controllerNSLabel set but no envInboundListenAddr env variable set.  In these cases we return an error which causes and error to be logged and the update is not sent.  I think we should do the same thing here.  If the inbound listen port cannot be found in the ExternalWorkload spec, we should return an error.

Additionally, we might want to add server side validation (validating webhook) that ensures that every ExternalWorkload has an inbound listen port configured.  Either that or a Linkerd check.

Reply:
> Additionally, we might want to add server side validation (validating webhook) that ensures that every ExternalWorkload has an inbound listen port configured. Either that or a Linkerd check.

That can come in a follow up you think?